### PR TITLE
TRUS-473 - Agent Internal Reference Number

### PR DESCRIPTION
**Pull request by codebysachin**

### CODE REVIEW

File: conf/messages.en [hunk 1]
```diff
@@ -307,7 +307,7 @@ trusteesUkAddress.line2 = Address line 2 (optional)
 trusteesUkAddress.line3 = Address line 3 (optional)
 trusteesUkAddress.townOrCity = Town or city
 trusteesUkAddress.postcode = Postcode
-trusteesUkAddress.postcode.hint = For example AB1 1AB
+trusteesUkAddress.postcode.hint = For example AB1 1ABtrustee-nino
```
Comment (manishsw) [inline conf/messages.en]: @codebysachin Revert to original text.

File: app/navigation/Navigator.scala [hunk 2]
```diff
@@ -26,10 +26,20 @@ import models._
 @Singleton
 class Navigator @Inject()() {
 
+//  private val agentRoutes : Page => AffinityGroup => UserAnswers => Call = {
+//    case WhatIsTheUTRPage => af => ua => {
```
Comment (manishsw) [inline app/navigation/Navigator.scala]: @codebysachin  Remove commented code.

File: app/navigation/Navigator.scala [hunk 3]
```diff
@@ -171,4 +181,9 @@ class Navigator @Inject()() {
     case CheckMode =>
       checkRouteMap(page)
   }
+
+//  def nextPage(page : Page, mode: Mode, af :AffinityGroup) : UserAnswers => Call = mode match {
+//    case NormalMode =>
```
Comment (manishsw) [inline app/navigation/Navigator.scala]: @codebysachin  same as above.

File: app/controllers/AgentInternalReferenceController.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+      val preparedForm = request.userAnswers.get(AgentInternalReferencePage) match {
+        case None => form
+        case Some(value) => form.fill(value)
+      }
+
+      Ok(view(preparedForm, mode))
+  }
+
+  def onSubmit(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async {
```
Comment (adamconder) [inline app/controllers/AgentInternalReferenceController.scala]: We should also ensure they are `hasAgentAffinityGroup()` onSubmit as well.

File: app/navigation/Navigator.scala [hunk 5]
```diff
@@ -88,12 +90,12 @@ class Navigator @Inject()() {
...
   }
 
-  private def trustHaveAUTRRoute(answers: UserAnswers) = {
+  private def trustHaveAUTRRoute(answers: UserAnswers, af: AffinityGroup) = {
     val condition = (answers.get(TrustRegisteredOnlinePage), answers.get(TrustHaveAUTRPage))
 
     condition match {
       case (Some(false), Some(true)) => routes.WhatIsTheUTRController.onPageLoad(NormalMode)
-      case (Some(false), Some(false)) => routes.TrustNameController.onPageLoad(NormalMode)
+      case (Some(false), Some(false)) => if(af == AffinityGroup.Organisation){ routes.AgentInternalReferenceController.onPageLoad(NormalMode)} else {routes.AgentInternalReferenceController.onPageLoad(NormalMode)}
```
Comment (adamconder) [inline app/navigation/Navigator.scala]: Your conditional expression goes to the same page here? Surely one of them should be different?

File: app/pages/AgentInternalReferencePage.scala [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package pages
+
+import play.api.libs.json.JsPath
+
+case object AgentInternalReferencePage extends QuestionPage[String] {
+
+  override def path: JsPath = JsPath \ toString
```
Comment (adamconder) [inline app/pages/AgentInternalReferencePage.scala]: Can we scope this into an Agent object so the path will be `override def path: JsPath = JsPath \ Agent \ toString` where `Agent` is an object with its own path.

File: conf/messages.en [hunk 7]
```diff
@@ -324,6 +324,15 @@ trusteesUkAddress.error.line3.invalidCharacters = Address line 3 must only inclu
 trusteesUkAddress.error.townOrCity.invalidCharacters = Town or city must only include letters a to z, numbers, ampersands (&), apostrophes, forward slashes, full stops, hyphens, round brackets and spaces
 trusteesUkAddress.error.postcode.invalidCharacters = Enter a real postcode
 
+agentInternalReference.title = What is your internal reference for this client?
+agentInternalReference.heading = What is your internal reference for this client?
+agentInternalReference.checkYourAnswersLabel = What is your internal reference for this client?
+agentInternalReference.paragraph1 = This reference will help you identify this trust in your saved registrations.
+agentInternalReference.hint = It must only include letters a to z, numbers 0 to 9, ampersands (&), apostrophes, grave accents (`) and hyphens.
+agentInternalReference.error.required = Enter your internal reference for this client.
```
Comment (adamconder) [inline conf/messages.en]: I don't think we require a full stop here.

File: conf/messages.en [hunk 8]
```diff
@@ -324,6 +324,15 @@ trusteesUkAddress.error.line3.invalidCharacters = Address line 3 must only inclu
...
+agentInternalReference.error.length = Your internal reference for this client must be 56 characters or less.
+agentInternalReference.error.invalidFormat = The internal reference for this client must only include letters a to z, numbers 0 to 9, ampersands (&), apostrophes, grave accents (`) and hyphens.
```
Comment (adamconder) [inline conf/messages.en]: No full stop required.

File: conf/messages.en [hunk 9]
```diff
@@ -324,6 +324,15 @@ trusteesUkAddress.error.line3.invalidCharacters = Address line 3 must only inclu
 trusteesUkAddress.error.townOrCity.invalidCharacters = Town or city must only include letters a to z, numbers, ampersands (&), apostrophes, forward slashes, full stops, hyphens, round brackets and spaces
 trusteesUkAddress.error.postcode.invalidCharacters = Enter a real postcode
 
+agentInternalReference.title = What is your internal reference for this client?
+agentInternalReference.heading = What is your internal reference for this client?
+agentInternalReference.checkYourAnswersLabel = What is your internal reference for this client?
+agentInternalReference.paragraph1 = This reference will help you identify this trust in your saved registrations.
+agentInternalReference.hint = It must only include letters a to z, numbers 0 to 9, ampersands (&), apostrophes, grave accents (`) and hyphens.
+agentInternalReference.error.required = Enter your internal reference for this client.
+agentInternalReference.error.length = Your internal reference for this client must be 56 characters or less.
```
Comment (adamconder) [inline conf/messages.en]: No full stop required.

File: test/navigation/navigators/MatchingRoutes.scala [hunk 10]
```diff
@@ -57,15 +58,34 @@ trait MatchingRoutes {
...
+        "user is an organisation" must {
 
-              navigator.nextPage(WhatIsTheUTRPage, NormalMode)(userAnswers)
-                .mustBe(routes.TrustNameController.onPageLoad(NormalMode))
+          "go to TrustName from WhatIsTheUTR" in {
+            forAll(arbitrary[UserAnswers]) {
+              userAnswers =>
+
+                navigator.nextPage(WhatIsTheUTRPage, NormalMode)(userAnswers)
+                  .mustBe(routes.TrustNameController.onPageLoad(NormalMode))
```
Comment (adamconder) [inline test/navigation/navigators/MatchingRoutes.scala]: This test should be failing as you've got in the implementation to go to AgentClientReference.

**Comment by codebysachin** [comment]
Changes made please review again.

**Comment by jassalrichy** [comment]
Looks good to go.

**Pull request merged**